### PR TITLE
Decouple peer-processing-logic from block-connection-logic (#3)

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Yet another step towards splitting main.cpp. After this and #8930 (which I will rebase after this is merged to clean it up a ton), we're ready to do the big code moves :).

This PR focuses on mapBlockSource, removing references to it from ProcessNewBlock and moving it to callsites, cleaning up logic in the process.

Built on #9026 to avoid conflicts and simplify the backporting (#9048).

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5885,27 +5889,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
-            CValidationState state;
             bool fNewBlock = false;
             // Since we requested this block (it was in mapBlocksInFlight), force it to be processed,
             // even if it would not be a candidate for new tip (missing previous block, chain not long enough, etc)
-            // BIP 152 permits peers to relay compact blocks after validating
-            // the header only; we should not punish peers if the block turns
-            // out to be invalid.
-            ProcessNewBlock(state, chainparams, &block, true, NULL, &fNewBlock);
-            int nDoS;
-            if (state.IsInvalid(nDoS)) {
```
Comment (sipa) [inline src/main.cpp]: So glad to see this duplication go!

**Review by sipa — COMMENTED** [review]
Concept ACK

File: src/main.cpp [hunk 2]
```diff
@@ -2976,8 +2976,6 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
             if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : NULL, txConflicted, txChanged)) {
                 if (state.IsInvalid()) {
                     // The block violates a consensus rule.
-                    if (!state.CorruptionPossible())
```
Comment (TheBlueMatt) [inline src/main.cpp]: Eek! Youre right. Just removed the commit...it wasnt important anyway.

File: src/main.cpp [hunk 3]
```diff
@@ -6068,13 +6072,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
+        const uint256 hash(block.GetHash());
         {
             LOCK(cs_main);
             // Also always process if we requested the block explicitly, as we may
             // need it even though it is not a candidate for a new best tip.
-            forceProcessing |= MarkBlockAsReceived(block.GetHash());
+            forceProcessing |= MarkBlockAsReceived(hash);
+            // mapBlockSource is only used for sending reject messages and DoS scores,
+            // so the race between here and cs_main in ProcessNewBlock is fine.
+            if (!mapBlockSource.count(hash))
```
Comment (TheBlueMatt) [inline src/main.cpp]: Fixed

**Comment by sipa** [comment]
Needs rebase.

**Comment by TheBlueMatt** [comment]
Rebased.

File: src/main.cpp [hunk 4]
```diff
@@ -5914,22 +5910,29 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 // updated, reject messages go out, etc.
                 MarkBlockAsReceived(resp.blockhash); // it is now an empty pointer
                 fBlockRead = true;
+                // mapBlockSource is only used for sending reject messages and DoS scores,
```
Comment (theuni) [inline src/main.cpp]: Stale comment? Or am I not seeing the race?

Comment (TheBlueMatt) [inline src/main.cpp]: Race if multiple nodes send you something near the same time you might send the DoS/Reject to the "wrong" node, of course we dont really care because we'll get to them next time they send us a block (and someone will still get it).

**Comment by TheBlueMatt** [comment]
As mentioned on IRC, this now reverts #9087, which conflicted, but no one registered any serious desire to maintain its functionality since it was realtively trivial.

File: src/main.cpp (hunk 4 above)
Comment (sipa) [inline src/main.cpp]: I assumed this just referred to the fact that we drop the cs_main lock between here and the call to `ProcessNewBlock` below, so in theory mapBlockSource could be modified in between, and we'd lose the expected callbacks.

**Comment by sipa** [comment]
utACK ae223576077448bd4ec250c73f5d8fe5e9a9ac7d

**Comment by theuni** [comment]
utACK ae223576077448bd4ec250c73f5d8fe5e9a9ac7d other than the comment nit. Obviously not a blocker though.

File: src/rpc/mining.cpp [hunk 5]
```diff
@@ -754,24 +753,19 @@ UniValue submitblock(const JSONRPCRequest& request)
...
         return "duplicate";
     }
-    if (fAccepted)
-    {
-        if (!sc.found)
-            return "inconclusive";
-        state = sc.state;
-    }
-    return BIP22ValidationResult(state);
+    if (!sc.found)
```
Comment (jtimon) [inline src/rpc/mining.cpp]: not sure why this shouldn't be ```if (fAccepted && !sc.found)```

File: src/main.cpp [hunk 6]
```diff
@@ -3787,22 +3787,26 @@ static bool AcceptBlock(const CBlock& block, CValidationState& state, const CCha
...
-        if (!ret)
+        if (!ret) {
+            GetMainSignals().BlockChecked(*pblock, state);
             return error("%s: AcceptBlock FAILED", __func__);
+        }
     }
 
     NotifyHeaderTip();
 
+    CValidationState state; // Only used to report errors, not invalidity - ignore it
```
Comment (jtimon) [inline src/main.cpp]: you don't need to duplicate  CValidationState state, you can move the first declaration out of the locked context.

File: src/main.cpp [hunk 7]
```diff
@@ -3787,19 +3787,15 @@ static bool AcceptBlock(const CBlock& block, CValidationState& state, const CCha
...
 
         // Store to disk
         CBlockIndex *pindex = NULL;
-        bool fNewBlock = false;
-        bool ret = AcceptBlock(*pblock, state, chainparams, &pindex, fForceProcessing, dbp, &fNewBlock);
-        if (pindex && pfrom) {
-            mapBlockSource[pindex->GetBlockHash()] = std::make_pair(pfrom->GetId(), fMayBanPeerIfInvalid);
-            if (fNewBlock) pfrom->nLastBlockTime = GetTime();
-        }
+        if (fNewBlock) *fNewBlock = false;
```
Comment (jtimon) [inline src/main.cpp]: s/if (fNewBlock) *fNewBlock = false;/*fNewBlock = false;/ ?

**Comment by jtimon** [comment]
ut ACK ae22357 apart from the couple of small nits.

**Comment by sdaftuar** [comment]
utACK ae223576077448bd4ec250c73f5d8fe5e9a9ac7d

File: src/main.cpp (hunk 6 above)
Comment (sdaftuar) [inline src/main.cpp]: Personally I like making it clear that this CValidationState is unrelated to the other one; it was confusing in the old code that validation errors from different blocks might get put in the same CValidationState (which would get cleared out in ActivateBestChainStep).

File: src/rpc/mining.cpp (hunk 5 above)
Comment (sipa) [inline src/rpc/mining.cpp]: If it is not accepted, there will certainly have been a callback to signal what was invalid about it. Thus, not having had a signal implies it was accepted.

File: src/main.cpp (hunk 6 above)
Comment (jtimon) [inline src/main.cpp]: Not a big deal anyway.

File: src/rpc/mining.cpp (hunk 5 above)
Comment (TheBlueMatt) [inline src/rpc/mining.cpp]: Technically this is a change in behavior...previously: if you got an error (disk space, etc) during connection you would print it in the BIP22 result function, after this patch (though, indeed, adding an fAccepted check wouldnt help) you will no longer print that, and your node will just exit (probably after retuning "success" from this function.

**Comment by rebroad** [comment]
Is this change part of a roadmap that is documented somewhere?

**Comment by jtimon** [comment]
@rebroad It is part of a longer branch by @TheBlueMatt : https://github.com/TheBlueMatt/bitcoin/tree/net_processing_file which does enough decoupling to cleanly separate the p2p messages part out of main.o. But I think the PRs (and commits) stand on their own (I did less review on the previous steps). Perhaps @TheBlueMatt can give more details.

**Comment by rebroad** [comment]
@jtimon thanks. I also meant to ask: why is this being done? Is it because main.cpp is considered to be too large?

**Comment by sipa** [comment]
@rebroad Splitting main into several files would be trivial to do, if we don't care much about clean separation between the resulting files. This PR is part of an ongoing effort to separate network processing and block validation with clear interfaces between them. The end result is not only splitting main into two files, but splitting it into two files that each have clearly-separated responsibilities.
